cachi2/core/package_managers/yarn_classic/workspaces.py

```diff
@@ -21,3 +21,86 @@ def _ensure_package_is_named(cls, package_contents: dict) -> dict:
         if "name" not in package_contents:
             raise ValueError("Workspaces must contain 'name' field.")
         return package_contents
+
+
+def ensure_no_path_leads_out(
+    paths: Iterable[Path],
+    source_dir: RootedPath,
+) -> None:
+    """Ensure no path leads out of source directory.
+
+    Raises an exception when any path is not relative to source directory.
+    Does nothing when path does not exist in the file system.
+    """
+    for path in paths:
+        try:
+            source_dir.join_within_root(path)
```

Comment:
There may be a subtle path-handling issue here. The source_dir is a RootedPath where the root is the root of the source git repository. In this case, I believe we want to be testing against a RootedPath where the root is the package_path (which I believe is the [workspace root](https://classic.yarnpkg.com/lang/en/docs/workspaces/#toc-limitations-caveats) referenced in the docs).

We may want to pass the package_root to extract_workspace_metadata rather than the input package and source_dir. We could then use the re_root method of the RootedPath class to set the appropriate root for the check.

Something like this:
```python
def extract_workspace_metadata(package_path: RootedPath) -> list[Workspace]:
    """Extract workspace metadata from a package."""
    processed_package = _read_package_from(package_path)
    workspaces_globs = _extract_workspaces_globs(processed_package)
    workspaces_paths = _get_workspace_paths(workspaces_globs, package_path)
    ensure_no_path_leads_out(workspaces_paths, package_path.re_root(package_path.subpath_from_root))
    _ensure_workspaces_are_well_formed(workspaces_paths)
    parsed_workspaces = []
    for wp in workspaces_paths:
        parsed_workspaces.append(
            Workspace(
                path=wp,
                package_contents=_read_package_from(package_path.join_within_root(wp)),
            )
        )
    return parsed_workspaces
```

Reply:
Not a blocker for me, since this is not really a risk to cachi2 and more of a Yarn install/build-time issue